test/project_types/extension/features/argo_serve_test.rb

```diff
@@ -43,24 +45,81 @@ def test_argo_serve_abort_when_server_start_failed
         )
       end
 
+      def test_forwards_resource_url
+        ShopifyCli::Tasks::EnsureDevStore.stubs(:call)
+        ShopifyCli::Tasks::EnsureEnv.stubs(:call)
+        ExtensionProject.stubs(:update_env_file)
+
+        project = ExtensionProject.current
+        def project.resource_url
+          "/test"
+        end
```

Comment:
`test_builds_resource_url_if_necessary` is currently failing CI with `Expected: "/generated", Actual: "/test"`.
This makes me think that the tests aren't independent from one another.

Is there a reason why we chose to `def` the `resource_url` method, instead of stubbing it?

Reply:
I replaced it with an `any_instance` stub now. Stubbing a concrete instance doesn't work as we force reload the project in the latest iteration of the code. Again, this stems from poor implementation of projects as singletons and shady environment handling. IMO, this should be evolved to:

```ruby
process = Pipeline.new(Task1.new, Task2.new, ...)
orchestrator = Orchestrator.new
orchestrator.load_env("./.env")
orchestrator.run_process(context, process)
orchestrator.save_end("./.env")
```

In reality, I would likely extract the actual marshalling and unmarshalling of the environment into its own objects, but this illustrates my point. We need clear layers and easy ways to compose complex process from simple pieces. Today we do all of this adhoc. And in many cases, the individual fragments share very little API surface.